### PR TITLE
Prs 453 fix subaccounts feature

### PR DESCRIPTION
**Pull request by s-medvedev**
#### Problem
Porting the subaccounts conflicts with `direct_account_pointers_in_program_input` and `syscall_parameter_address_restrictions` features. This change resolves conflicts.

#### Summary of Changes
1. Reserve subaccounts slots regardless of `direct_account_pointers_in_program_input` status
2. Enable MM_INPUT_REGION for subaccounts inside `CallerAccount::from_account_info()` and `CallerAccount::from_sol_account_info()` because AccountView for SubaccountSlot was placed inside this memory region, and only the core passes these addresses into the function.

### CODE REVIEW

File: program-runtime/src/serialization.rs [hunk 1]
```diff
@@ -700,82 +699,72 @@ fn serialize_parameters_for_abiv1(
     s.write::<u64>((instruction_data.len() as u64).to_le());
     let instruction_data_offset = s.write_all(instruction_data);
     s.write_all(program_id.as_ref());
+    let align_offset = (instruction_data.len() as *const u8).align_offset(BPF_ALIGN_OF_U128);
```
Comment (BoogVAr) [inline program-runtime/src/serialization.rs]: **HIGH (pending verification) — pointer-array alignment basis.** On the base branch the SIMD-0449 pointer array was padded by the running position (`(size as *const u8).align_offset(...)`), which aligns it to 16 for any account set. This changes the basis to `instruction_data.len()`.

Because a **duplicate** account block is 8 bytes (regular blocks are multiples of 16, and `accounts.len()`+`data_len` = 16), the padded offset is `≡ 8*ndup (mod 16)`: 0 for an even number of duplicate accounts, **8 for an odd number**. So with an odd duplicate count the pointer array lands 8 bytes off the base boundary.

The only test uses `[1,1,2,3,4,4,5,6]` (2 dups) and deserializes the inline layout, so this isn't caught. Please align the pointer array by actual position for the `direct_account_pointers_program_input` branch (as the base did), or show the two bases are equivalent for all account sets.

File: program-runtime/src/serialization.rs [hunk 2]
```diff
@@ -632,29 +632,28 @@ fn serialize_parameters_for_abiv1(
     }
     size += size_of::<u64>() // data len
     + instruction_data.len()
-    + size_of::<Pubkey>(); // program id;
+    + size_of::<Pubkey>() // program id
+    // F10: alignment padding so the u64 subaccount count lands on BPF_ALIGN_OF_U128
```
Comment (BoogVAr) [inline program-runtime/src/serialization.rs]: Nit (docs): this comment says the padding aligns "the u64 subaccount count", but no subaccount-count `u64` is written here — the padding precedes the (optional) pointer array and the view-buffer region. Suggest rewording to describe what actually follows.

File: program-runtime/src/serialization.rs [hunk 3]
```diff
@@ -700,82 +699,72 @@ fn serialize_parameters_for_abiv1(
...
+    //      immediately followed by the 88-byte serialized slot header
+    //      (NON_DUP_MARKER + flags + key + owner + lamports + data_len),
+    //   2. an empty readonly placeholder data region whose VM address is
+    //      reserved with SUBACCOUNT_SLOT_DATA_RESERVED_VM_BYTES so a later
+    //      MemoryMapping::replace_region can install a region backed by the
+    //      loaded subaccount's AccountSharedData without colliding with the
+    //      next slot's reserved range.
+    // Close any preceding region first so each slot's view-buffer + header
+    // pair lives in its own region.
+    s.push_region();
```
Comment (BoogVAr) [inline program-runtime/src/serialization.rs:722]: Note (design intent): with `direct_account_pointers_program_input` active, `sol_load_subaccount` is gated off, yet this now serializes all 16 slots (~2.3 KB) and pushes 32 extra memory regions (16 `push_region` + 16 `push_data_placeholder`) on every invocation. That matches the PR's stated goal (reserve regardless), so not a blocker — just flagging the per-CPI region-count / CU cost in the SIMD-0449 configuration in case it matters on hot paths.

File: program-runtime/src/cpi.rs [hunk 4]
```diff
@@ -307,8 +307,10 @@ impl<'a> CallerAccount<'a> {
         _vm_addr: u64,
         account_info: &solana_account_info::AccountInfo,
         account_metadata: &crate::invoke_context::SerializedAccountMetadata,
+        enable_mm_input: bool,
     ) -> Result<CallerAccount<'a>, Error> {
         use crate::memory::{translate_type, translate_type_mut_for_cpi};
+        let max_ptr_value = solana_sbpf::ebpf::MM_INPUT_START + if enable_mm_input {solana_sbpf::ebpf::MM_REGION_SIZE} else {0};
```
Comment (BoogVAr) [inline program-runtime/src/cpi.rs:313]: Security review of the guard relaxation: this looks sound. When `enable_mm_input=true` the outer `>= MM_INPUT_START` guards widen to `>= MM_INPUT_START + MM_REGION_SIZE`, but the inner `check_account_info_pointer` still strict-equality-locks `lamports`/`data`/`key`/`owner` to the runtime slot metadata, the `AccountView` itself is read from the runtime-controlled `view_addr`, and `vm_len_addr` is derived from the view struct layout — so I don't see a new pointer-redirect primitive. Recommend adding a negative test: a program that sets a subaccount `AccountInfo` lamports/data pointer to another account/slot in the input region should be rejected with `InvalidPointer`. The normal path keeps `enable_mm_input=false`, so it's unchanged.

**Review by BoogVAr — COMMENTED** [review]
## Code review — subaccounts × v4 ABI feature conflict

Thanks — the approach (reserve the subaccount region unconditionally, and relax the `syscall_parameter_address_restrictions` pointer guard only for subaccount views via `enable_mm_input`) reads cleanly and the scope is tight. I verified the size/write symmetry: both sides use the same `align_offset` basis and the same conditional `accounts.len()*8`, so the pre-allocated buffer is exactly filled — no overflow. The `enable_mm_input=false` default keeps the normal CPI path bit-identical to before.

One concern I'd like resolved before merge, plus a few smaller notes (inline).

### 🔴 HIGH (pending verification) — pointer-array alignment basis changed for the SIMD-0449 path

On the base branch the account-pointer array was padded with `(size as *const u8).align_offset(BPF_ALIGN_OF_U128)` — i.e. by the **running serialized position** — which aligns the array to 16 for any account set. This PR switches both paths to `(instruction_data.len() as *const u8).align_offset(...)`.

Working through the layout: each regular account block is a multiple of 16 bytes, `accounts.len()` + `data_len` u64s are 16 bytes, but each **duplicate** account block is 8 bytes. So the offset at the end of the `instruction_data.len()`-based padding is `≡ 8*ndup (mod 16)` — i.e. 0 when the number of duplicate accounts is even, **8 when it is odd**. With an odd duplicate count the pointer array (SIMD-0449) starts 8 bytes off the boundary the base layout produced.

The existing `test_serialize_parameters` uses `[1,1,2,3,4,4,5,6]` (2 duplicates — even), and its `deserialize` reads the inline layout rather than the pointer array, so this case is not exercised. Whether the 8-byte shift actually breaks a program depends on the on-chain SIMD-0449 pointer-array deserializer (not in this diff) — so I've flagged it PLAUSIBLE, but the divergence from the base layout is confirmed by the diff.

**Suggested fix:** for the `direct_account_pointers_program_input` branch, align the pointer array by the actual serialized position (as the base did), or demonstrate the two bases are equivalent for all account sets — then add a test with `direct_account_pointers=true` + subaccounts + an **odd** duplicate-account count.

### Smaller notes (inline)
- Test coverage gap for the direct-pointers + subaccount + odd-duplicate case.
- Security: the pointer-guard relaxation looks sound (inner `check_account_info_pointer` strict-equality-locks lamports/data/key/owner to runtime slot metadata, outer pointers are bound to the runtime `view_addr`) — but there is no negative test proving a program cannot redirect a subaccount pointer into another account within the input region.
- Stale comment referencing a "u64 subaccount count" that isn't written.
- Under SIMD-0449 the 16 slots (~2.3 KB + 32 memory regions) are serialized on every invocation though `sol_load_subaccount` is gated off — accepted design intent, noting the per-CPI region/CU cost.

Also confirmed: no dangling line-continuation in `ci/parasol-node-run.sh` after the removal, and no AI-authorship attribution anywhere in the diff/commits/description.

File: program-runtime/src/cpi.rs [hunk 5]
```diff
@@ -307,8 +307,10 @@ impl<'a> CallerAccount<'a> {
...
```
[the remark this answers is not in the document]
Comment (s-medvedev) [reply program-runtime/src/cpi.rs:314]: Fixed

File: program-runtime/src/serialization.rs [hunk 6]
```diff
@@ -700,82 +699,72 @@ fn serialize_parameters_for_abiv1(
     s.write::<u64>((instruction_data.len() as u64).to_le());
     let instruction_data_offset = s.write_all(instruction_data);
     s.write_all(program_id.as_ref());
+    let align_offset = (instruction_data.len() as *const u8).align_offset(BPF_ALIGN_OF_U128);
+    s.fill_write(align_offset, 0)
+        .map_err(|_| InstructionError::InvalidArgument)?;
```
[the remark this answers is not in the document]
Comment (s-medvedev) [reply program-runtime/src/serialization.rs:700]: Fixed

File: program-runtime/src/serialization.rs [hunk 7]
```diff
@@ -632,29 +632,28 @@ fn serialize_parameters_for_abiv1(
...
     };
 
+    // F10 (SIMD-0449 mutual-gate): when direct account pointers are NOT
+    // active, reserve the subaccount slot region in the same post-program_id
+    // byte range the pointer array would occupy. The two are mutually
+    // exclusive by construction — `sol_load_subaccount` is gated off whenever
+    // SIMD-0449 is active — so they never collide. In the direct-map model no
+    // real-subaccount records are serialized (data is direct-mapped); only the
+    // MAX_SUBACCOUNT_SLOTS reserved slots (each an account-view buffer + the
+    // 88-byte header) are laid out, after alignment padding.
```
[the remark this answers is not in the document]
Comment (s-medvedev) [reply program-runtime/src/serialization.rs]: Fixed

File: program-runtime/src/serialization.rs (hunk 1 above)
Comment (s-medvedev) [reply program-runtime/src/serialization.rs]: Fixed

File: program-runtime/src/serialization.rs (hunk 2 above)
Comment (s-medvedev) [reply program-runtime/src/serialization.rs]: Fixed

File: program-runtime/src/serialization.rs (hunk 3 above)
Comment (s-medvedev) [reply program-runtime/src/serialization.rs:722]: Won't fix

File: program-runtime/src/cpi.rs (hunk 4 above)
Comment (s-medvedev) [reply program-runtime/src/cpi.rs:313]: Won't fix
